### ISSUE TITLE
TextMergeViewer deadlock 

### ISSUE THREAD

**Issue by jukzi** [eclipse-platform/eclipse.platform#282]
My eclipse IDE just stuck. I had a git pull which reported a change in "build.properties" and "MANIFEST.MF". I opened compare viewer and used "replace with" in Git stating view for both files.
Swt Thread "main" waits for a worker thread that contrary waits that something in SWT finished = deadlock:

2022-11-23 11:53:26
Full thread dump OpenJDK 64-Bit Server VM (17.0.5+8 mixed mode, sharing):

```
"main" #1 prio=6 os_prio=0 cpu=76359.38ms elapsed=3218.47s tid=0x000001f661ecc910 nid=0x590c waiting for monitor entry  [0x0000003e654fd000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at org.eclipse.core.internal.filebuffers.SynchronizableDocument.getModificationStamp(SynchronizableDocument.java:151)
        - waiting to lock <0x000000075e6b4548> (a java.lang.Object)
        at org.eclipse.text.undo.DocumentUndoManager$UndoableTextChange.canUndo(DocumentUndoManager.java:163)
        at org.eclipse.core.commands.operations.DefaultOperationHistory.canUndo(DefaultOperationHistory.java:269)
        at org.eclipse.text.undo.DocumentUndoManager.undoable(DocumentUndoManager.java:835)
        at org.eclipse.jface.text.TextViewerUndoManager.undoable(TextViewerUndoManager.java:348)
        at org.eclipse.jface.text.TextViewer.canDoOperation(TextViewer.java:3886)
        at org.eclipse.jface.text.source.SourceViewer.canDoOperation(SourceViewer.java:816)
        at org.eclipse.compare.internal.MergeSourceViewer$TextOperationAction.isEnabled(MergeSourceViewer.java:139)
        at org.eclipse.compare.internal.MergeSourceViewer$TextOperationAction.update(MergeSourceViewer.java:144)
        at org.eclipse.compare.internal.MergeSourceViewer.updateContentDependantActions(MergeSourceViewer.java:818)
        at org.eclipse.compare.internal.MergeSourceViewer$$Lambda$2457/0x00000008023456f0.run(Unknown Source)
        at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
        at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
        - locked <0x00000007b16f92c8> (a org.eclipse.swt.widgets.RunnableLock)
        at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4043)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3648)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
        at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
        at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
        at org.eclipse.ui.internal.Workbench$$Lambda$180/0x0000000800db70a0.run(Unknown Source)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
        at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
        at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
        at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
        at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17.0.5/Native Method)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17.0.5/NativeMethodAccessorImpl.java:77)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17.0.5/DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(java.base@17.0.5/Method.java:568)
        at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
        at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
        at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
        at org.eclipse.equinox.launcher.Main.main(Main.java:1440)

"Worker-33: Building" #350 prio=5 os_prio=0 cpu=22296.88ms elapsed=1838.26s tid=0x000001f65d401aa0 nid=0xcf4 waiting on condition  [0x0000003e6a7fd000]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.5/Native Method)
        - parking to wait for  <0x00000007b16530c8> (a java.util.concurrent.Semaphore$NonfairSync)
        at java.util.concurrent.locks.LockSupport.parkNanos(java.base@17.0.5/LockSupport.java:252)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.5/AbstractQueuedSynchronizer.java:717)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(java.base@17.0.5/AbstractQueuedSynchronizer.java:1074)
        at java.util.concurrent.Semaphore.tryAcquire(java.base@17.0.5/Semaphore.java:415)
        at org.eclipse.ui.internal.PendingSyncExec.acquire(PendingSyncExec.java:39)
        at org.eclipse.ui.internal.PendingSyncExec.waitUntilExecuted(PendingSyncExec.java:88)
        at org.eclipse.ui.internal.UISynchronizer.syncExec(UISynchronizer.java:142)
        at org.eclipse.swt.widgets.Display.syncExec(Display.java:4790)
        at org.eclipse.compare.internal.Utilities.firePropertyChange(Utilities.java:182)
        at org.eclipse.compare.internal.Utilities.firePropertyChange(Utilities.java:169)
        at org.eclipse.compare.contentmergeviewer.ContentMergeViewer.fireDirtyState(ContentMergeViewer.java:1181)
        at org.eclipse.compare.contentmergeviewer.ContentMergeViewer.setLeftDirty(ContentMergeViewer.java:1197)
        at org.eclipse.compare.contentmergeviewer.TextMergeViewer.updateDirtyState(TextMergeViewer.java:3290)
        at org.eclipse.compare.contentmergeviewer.TextMergeViewer$ContributorInfo.elementDirtyStateChanged(TextMergeViewer.java:1049)
        at org.eclipse.ui.editors.text.TextFileDocumentProvider$FileBufferListener.dirtyStateChanged(TextFileDocumentProvider.java:305)
        at org.eclipse.core.internal.filebuffers.TextFileBufferManager$3.run(TextFileBufferManager.java:670)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
        at org.eclipse.core.internal.filebuffers.TextFileBufferManager.fireDirtyStateChanged(TextFileBufferManager.java:667)
        at org.eclipse.core.internal.filebuffers.ResourceTextFileBuffer$DocumentListener.documentChanged(ResourceTextFileBuffer.java:82)
        at org.eclipse.jface.text.AbstractDocument.doFireDocumentChanged2(AbstractDocument.java:748)
        at org.eclipse.jface.text.AbstractDocument.doFireDocumentChanged(AbstractDocument.java:717)
        at org.eclipse.jface.text.AbstractDocument.doFireDocumentChanged(AbstractDocument.java:701)
        at org.eclipse.jface.text.AbstractDocument.fireDocumentChanged(AbstractDocument.java:775)
        at org.eclipse.jface.text.AbstractDocument.replace(AbstractDocument.java:1100)
        at org.eclipse.core.internal.filebuffers.SynchronizableDocument.replace(SynchronizableDocument.java:176)
        - locked <0x000000075e6b4548> (a java.lang.Object)
        at org.eclipse.jface.text.AbstractDocument.replace(AbstractDocument.java:1118)
        at org.eclipse.core.internal.filebuffers.SynchronizableDocument.replace(SynchronizableDocument.java:164)
        - locked <0x000000075e6b4548> (a java.lang.Object)
        at org.eclipse.text.edits.ReplaceEdit.performDocumentUpdating(ReplaceEdit.java:79)
        at org.eclipse.text.edits.TextEdit.traverseDocumentUpdating(TextEdit.java:920)
        at org.eclipse.text.edits.TextEdit.traverseDocumentUpdating(TextEdit.java:913)
        at org.eclipse.text.edits.TextEditProcessor.executeDo(TextEditProcessor.java:196)
        at org.eclipse.text.edits.TextEdit.dispatchPerformEdits(TextEdit.java:742)
        at org.eclipse.text.edits.TextEditProcessor.performEdits(TextEditProcessor.java:158)
        at org.eclipse.text.edits.TextEdit.apply(TextEdit.java:714)
        at org.eclipse.text.edits.TextEdit.apply(TextEdit.java:738)
        at org.eclipse.pde.internal.ui.util.PDEModelUtility.generateModelEdits(PDEModelUtility.java:342)
        at org.eclipse.pde.internal.ui.util.PDEModelUtility.modifyModel(PDEModelUtility.java:279)
        at org.eclipse.pde.ds.internal.annotations.DSAnnotationCompilationParticipant.updateProject(DSAnnotationCompilationParticipant.java:467)
        at org.eclipse.pde.ds.internal.annotations.DSAnnotationCompilationParticipant.buildFinished(DSAnnotationCompilationParticipant.java:411)
        at org.eclipse.jdt.internal.core.builder.JavaBuilder.build(JavaBuilder.java:251)
        at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:1020)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
        at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:247)
        at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:303)
        at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:392)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
        at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:395)
        at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:506)
        at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:454)
        at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:536)
        at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:161)
        at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:255)
        at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```
Already reported in:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=519002 [wontfix]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=319198 [wontfix]

**Comment by jukzi** [comment]
Can we just change syncExec to asyncExec in org.eclipse.compare.internal.Utilities.firePropertyChange()? (as suggested in  https://bugs.eclipse.org/bugs/show_bug.cgi?id=319198#c4)

**Comment by vogella** [comment]
> Can we just change syncExec to asyncExec in org.eclipse.compare.internal.Utilities.firePropertyChange()? (as suggested in https://bugs.eclipse.org/bugs/show_bug.cgi?id=319198#c4)

Would be nice to try that approach, any open compare editor seems to slow down the ide. I tend to close them frequently to avoid sluggish UI.

**Comment by jukzi** [comment]
I can not reproduce the Error, so i don't know if its fixed by this change :-(
I only remember the issue involved an open merge view on MANIFEST.MF, a conflicting change in git, enabled "DS-annotations" and a class that was changed by one of the commits which was annotated @Component - which would create a line Service-Component: "OSGI-INF/projecrt.MyClass.xml" in the MANIFEST.

**Comment by iloveeclipse** [comment]
reopening because #283 caused regression in #464 and will be reverted.

**Comment by iloveeclipse** [comment]
The problem can still be reproduced by using Compare editor & DS enabled plugin project, I saw it yesterday.
I had a diff editor opened on MANIFEST.MF (changed) by a DS builder and I was reverting changes created by DS builder from the staging view.

* The manifest was from a test project in `eclipse.pde.ui/ds/org.eclipse.pde.ds.annotations.tests/projects/test1`.
* Rename `test.project` to `.project`, import it as a project in Eclipse and see that MANIFEST.MF and few other files will be changed.
* Check the changes in the Compare editor.
* Revert changes from Staging view
* Kaboom. 

Stack from current master.

**Innocent main thread** just tries to "paint" code in Compare editor
```
"main" eclipse-platform/eclipse.platform#3 [1519803] prio=6 os_prio=0 cpu=975689,68ms elapsed=31030,54s tid=0x00007fe560030b00 nid=1519803 waiting for monitor entry  [0x00007fe565b1a000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.eclipse.core.internal.filebuffers.SynchronizableDocument.getLineInformation(SynchronizableDocument.java:384)
	- waiting to lock <0x0000000459aeecb0> (a java.lang.Object)
	at org.eclipse.jface.text.DefaultDocumentAdapter.doGetLine(DefaultDocumentAdapter.java:147)
	at org.eclipse.jface.text.DefaultDocumentAdapter.getLine(DefaultDocumentAdapter.java:160)
	at org.eclipse.swt.custom.StyledTextRenderer.makeLineDrawInfo(StyledTextRenderer.java:467)
	at org.eclipse.swt.custom.StyledTextRenderer.drawLines(StyledTextRenderer.java:520)
	at org.eclipse.swt.custom.StyledText.handlePaint(StyledText.java:5993)
	at org.eclipse.swt.custom.StyledText.lambda$32(StyledText.java:5487)
	at org.eclipse.swt.custom.StyledText$$Lambda/0x0000000037b8bc00.handleEvent(Unknown Source)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:91)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5893)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1656)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1682)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1665)
	at org.eclipse.swt.widgets.Control.gtk_draw(Control.java:3902)
	at org.eclipse.swt.widgets.Scrollable.gtk_draw(Scrollable.java:394)
	at org.eclipse.swt.widgets.Composite.gtk_draw(Composite.java:507)
	at org.eclipse.swt.widgets.Canvas.gtk_draw(Canvas.java:173)
	at org.eclipse.swt.widgets.Widget.windowProc(Widget.java:2612)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:6848)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6200)
	at org.eclipse.swt.internal.gtk3.GTK3.gtk_main_do_event(Native Method)
	at org.eclipse.swt.widgets.Display.eventProc(Display.java:1624)
	at org.eclipse.swt.internal.gtk3.GTK3.gtk_main_iteration_do(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4542)
```

**Bad builder** thread does many things wrong.
- `DSAnnotationCompilationParticipant` updates PDE model  (workspace lock during build)
- `PDEModelUtility` applies edits on `SynchronizableDocument` (document lock on document change)
- `ContentMergeViewer.fireDirtyState` calls `Display.syncExec()` (UI lock needed)
```
"Worker-646: Building" #8867 [2152209] prio=5 os_prio=0 cpu=672,49ms elapsed=1560,34s tid=0x00007fe42001c660 nid=2152209 waiting on condition  [0x00007fe07caf7000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@25/Native Method)
	- parking to wait for  <0x00000004689f14c0> (a java.util.concurrent.Semaphore$NonfairSync)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@25/LockSupport.java:271)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@25/AbstractQueuedSynchronizer.java:792)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(java.base@25/AbstractQueuedSynchronizer.java:1166)
	at java.util.concurrent.Semaphore.tryAcquire(java.base@25/Semaphore.java:415)
	at org.eclipse.ui.internal.PendingSyncExec.acquire(PendingSyncExec.java:39)
	at org.eclipse.ui.internal.PendingSyncExec.waitUntilExecuted(PendingSyncExec.java:88)
	at org.eclipse.ui.internal.UISynchronizer.syncExec(UISynchronizer.java:145)
	at org.eclipse.swt.widgets.Display.syncExec(Display.java:5998)
	at org.eclipse.compare.internal.Utilities.firePropertyChange(Utilities.java:184)
	at org.eclipse.compare.internal.Utilities.firePropertyChange(Utilities.java:170)
	at org.eclipse.compare.contentmergeviewer.ContentMergeViewer.fireDirtyState(ContentMergeViewer.java:1218)
	at org.eclipse.compare.contentmergeviewer.ContentMergeViewer.setRightDirty(ContentMergeViewer.java:1251)
	at org.eclipse.compare.contentmergeviewer.TextMergeViewer.updateDirtyState(TextMergeViewer.java:3417)
	at org.eclipse.compare.contentmergeviewer.TextMergeViewer$ContributorInfo.elementDirtyStateChanged(TextMergeViewer.java:1082)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider$FileBufferListener.dirtyStateChanged(TextFileDocumentProvider.java:339)
	at org.eclipse.core.internal.filebuffers.TextFileBufferManager$3.run(TextFileBufferManager.java:693)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.filebuffers.TextFileBufferManager.fireDirtyStateChanged(TextFileBufferManager.java:690)
	at org.eclipse.core.internal.filebuffers.ResourceTextFileBuffer$DocumentListener.documentChanged(ResourceTextFileBuffer.java:76)
	at org.eclipse.jface.text.AbstractDocument.doFireDocumentChanged2(AbstractDocument.java:762)
	at org.eclipse.jface.text.AbstractDocument.doFireDocumentChanged(AbstractDocument.java:730)
	at org.eclipse.jface.text.AbstractDocument.doFireDocumentChanged(AbstractDocument.java:714)
	at org.eclipse.jface.text.AbstractDocument.fireDocumentChanged(AbstractDocument.java:790)
	at org.eclipse.jface.text.AbstractDocument.replace(AbstractDocument.java:1132)
	at org.eclipse.core.internal.filebuffers.SynchronizableDocument.replace(SynchronizableDocument.java:271)
	- locked <0x0000000459aeecb0> (a java.lang.Object)
	at org.eclipse.jface.text.AbstractDocument.replace(AbstractDocument.java:1150)
	at org.eclipse.core.internal.filebuffers.SynchronizableDocument.replace(SynchronizableDocument.java:259)
	- locked <0x0000000459aeecb0> (a java.lang.Object)
	at org.eclipse.text.edits.InsertEdit.performDocumentUpdating(InsertEdit.java:78)
	at org.eclipse.text.edits.TextEdit.traverseDocumentUpdating(TextEdit.java:941)
	at org.eclipse.text.edits.TextEdit.traverseDocumentUpdating(TextEdit.java:933)
	at org.eclipse.text.edits.TextEditProcessor.executeDo(TextEditProcessor.java:201)
	at org.eclipse.text.edits.TextEdit.dispatchPerformEdits(TextEdit.java:753)
	at org.eclipse.text.edits.TextEditProcessor.performEdits(TextEditProcessor.java:160)
	at org.eclipse.text.edits.TextEdit.apply(TextEdit.java:725)
	at org.eclipse.text.edits.TextEdit.apply(TextEdit.java:749)
	at org.eclipse.pde.internal.ui.util.PDEModelUtility.generateModelEdits(PDEModelUtility.java:372)
	at org.eclipse.pde.internal.ui.util.PDEModelUtility.modifyModel(PDEModelUtility.java:304)
	at org.eclipse.pde.ds.internal.annotations.DSAnnotationCompilationParticipant.updateProject(DSAnnotationCompilationParticipant.java:458)
	at org.eclipse.pde.ds.internal.annotations.DSAnnotationCompilationParticipant.buildFinished(DSAnnotationCompilationParticipant.java:412)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.build(JavaBuilder.java:288)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:1109)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:299)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:355)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:449)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:452)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:564)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:512)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:594)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:208)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:309)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```


There are at at least three equally bad problems :-(

1) Looking at https://github.com/eclipse-platform/eclipse.platform/issues/464, it seems that Compare editor (`TextMergeViewer` & `ContentMergeViewer` & Co) originally "synchronized" diff operations via UI thread / `syncExec()`. If that synchronization is removed (which is a "natural" solution to fix this deadlock like it was done in https://github.com/eclipse-platform/eclipse.platform/pull/283), Compare editor starts misbehave to the point of being unusable. A proper solution to synchronize diff changes must be understood & `syncExec()` finally removed.
2) `DSAnnotationCompilationParticipant` during build triggers PDE code that modifies `SynchronizableDocument` (document lock!) while being called from a build job (resource lock!). That must be decoupled (acquiring two locks is always a bad idea) from the build.
3) `ResourceTextFileBuffer.DocumentListener.documentChanged(DocumentEvent)` is a listener on **Synchronizable**Document (so it knows it will be triggered with a lock held) but it nofifies its own listeners in the same thread.

For now it **seems** that solving point 3) is the easiest part (but so this was assumption in https://github.com/eclipse-platform/eclipse.platform/pull/283 too).

**Comment by iloveeclipse** [comment]
> For now it **seems** that solving point 3) is the easiest part (but so this was assumption in [#283](https://github.com/eclipse-platform/eclipse.platform/pull/283) too).

That was unfortunately also too risky, as it would affect lot of unrelated code. The particular deadlock has a local solution, see https://github.com/eclipse-pde/eclipse.pde/pull/2186.

General problem (number 4) arises from `SynchronizableDocument` implementation, which should never be implemented in a way it was done, where notifications are sent to listeners while holding a lock.

**Comment by iloveeclipse** [comment]
> General problem (number 4) arises from `SynchronizableDocument` implementation, which should never be implemented in a way it was done, where notifications are sent to listeners while holding a lock.

Follow up: https://github.com/eclipse-platform/eclipse.platform.ui/issues/3685

**Comment by laeubi** [comment]
> > General problem (number 4) arises from `SynchronizableDocument` implementation, which should never be implemented in a way it was done, where notifications are sent to listeners while holding a lock.
> 
> Follow up: [eclipse-platform/eclipse.platform.ui#3685](https://github.com/eclipse-platform/eclipse.platform.ui/issues/3685)

As elaborated here:
- https://github.com/eclipse-platform/eclipse.platform.ui/issues/3685#issuecomment-3768630055

I think this is by intention as otherwise the listener contract would break.

Just assume you are notified that a document is `documentAboutToBeChanged`... if this would happen outside synchronization, then the document is likely already changed when you get the event (and even worse you possibly getting a reentrant notification for `documentChanged`) and even leading to multiple threads calling the listener in any order and all the document methods will throw `BadLocationException` all the time.

One could possibly have this implemented with a read/write lock but this would not work with an arbitrary object what seems to be wanted for some case.
